lib/segment/src/index/field_index/binary_index.rs

```diff
@@ -0,0 +1,521 @@
+use std::sync::Arc;
+
+use bitflags::bitflags;
+use bitvec::prelude::*;
+use parking_lot::RwLock;
+use rocksdb::DB;
+
+use super::{CardinalityEstimation, PayloadFieldIndex, PrimaryCondition, ValueIndexer};
+use crate::common::rocksdb_wrapper::DatabaseColumnWrapper;
+use crate::entry::entry_point::{OperationError, OperationResult};
+use crate::telemetry::PayloadIndexTelemetry;
+use crate::types::{
+    FieldCondition, Match, MatchValue, PayloadKeyType, PointOffsetType, ValueVariants,
+};
+
+bitflags! {
+    /// Due to being able to store multi-values, the binary index is not a simple
+    /// bitset, but rather a pair of bitsets, one for true values and one for false values.
+    pub struct BinaryItem: u8 {
+        const TRUE = 0b00000001;
+        const FALSE = 0b00000010;
+    }
+}
+
+impl BinaryItem {
+    pub fn from_bools(has_true: bool, has_false: bool) -> Self {
+        let mut item = Self::empty();
+        item.set(Self::TRUE, has_true);
+        item.set(Self::FALSE, has_false);
+        item
+    }
+}
+
+impl From<bool> for BinaryItem {
+    fn from(src: bool) -> Self {
+        if src {
+            BinaryItem::TRUE
+        } else {
+            BinaryItem::FALSE
+        }
+    }
+}
+
+struct BinaryMemory {
+    trues: BitVec,
+    falses: BitVec,
+    trues_count: usize,
+    falses_count: usize,
+    indexed_count: usize,
+}
+
+impl BinaryMemory {
+    pub fn new() -> Self {
+        Self {
+            trues: BitVec::new(),
+            falses: BitVec::new(),
+            trues_count: 0,
+            falses_count: 0,
+            indexed_count: 0,
+        }
+    }
+
+    pub fn get(&self, id: PointOffsetType) -> BinaryItem {
+        debug_assert!(self.trues.len() == self.falses.len());
+        if (id as usize) >= self.trues.len() {
+            return BinaryItem::empty();
+        }
+
+        unsafe {
+            // SAFETY: we just checked that the id is within bounds
+            let has_true = *self.trues.get_unchecked(id as usize).as_ref();
+            let has_false = *self.falses.get_unchecked(id as usize).as_ref();
+            BinaryItem::from_bools(has_true, has_false)
+        }
+    }
+
+    pub fn set_or_insert(&mut self, id: PointOffsetType, item: BinaryItem) {
```

Comment:
This only allows setting true/false bits to `1`. I wonder if this should also set bits back to `0`, or did you have a good reason to do it this way?

Setting both `1` and `0` in a single function would probably simplify things, and would _remove_ the need for `remove`. I believe we always know the true/false states of a point when setting.

Reply:
Whenever we insert a point to the index, it is always removed first, which is why `remove` is part of the trait. That said, the semantics of `set` imply changing to whatever is in the item, so we should indeed also switch back to 0